### PR TITLE
asyncio-style queue implementations

### PR DESCRIPTION
**Pull request by ktbarrett**
Identical to and supersedes #2297. Adds type annotations and documentation. Pinging @alexforencich.

### CODE REVIEW

File: tests/test_cases/test_cocotb/test_queues.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    assert not q.empty()
+    assert q.full()
+
+    # overflow
+    got_exception = False
+    try:
+        q.put_nowait(100)
+    except QueueFull:
+        got_exception = True
+    assert got_exception
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_queues.py:52]: I'm thinking I should replace this with `pytest.raises` instead.

File: tests/test_cases/test_cocotb/test_queues.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    getter_list = []
+
+    # test put contention
+    for k in range(NUM_PUTTERS):
+        cr_list.append(cocotb.fork(putter(putter_list, k)))
+
+    assert q.qsize() == QUEUE_SIZE
+
+    # test killed putter
+    cr = cocotb.fork(putter(putter_list, 100))
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_queues.py:122]: Sure

File: cocotb/queue.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            result += ' tasks={}'.format(self._unfinished_tasks)
+        return result
+
+    def qsize(self) -> int:
+        """Number of items in the queue."""
+        return len(self._queue)
+
+    @property
+    def maxsize(self) -> int:
+        """Number of items allowed in the queue."""
```
Comment (cmarqu) [inline cocotb/queue.py:80]: Not sure if the "infinite if <= 0" behavior should be mentioned here once more?

File: cocotb/queue.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            raise QueueFull()
+        self._put(item)
+        self._unfinished_tasks += 1
+        self._finished.clear()
+        self._wakeup_next(self._getters)
+
+    async def get(self) -> T:
+        """Remove and return an item from the queue.
+
+        If queue is empty, wait until an item is available.
```
Comment (cmarqu) [inline cocotb/queue.py:126]:
```suggestion
        If the queue is empty, wait until an item is available.
```

**Comment by marlonjames** [comment]
Could you improve the tests to get full coverage?

File: tests/test_cases/test_cocotb/test_queues.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    # remove value that forced contention
+    assert q.get_nowait() is None, "Popped unexpected value"
+
+    # test fair scheduling by ensuring that each putter is serviced for its first
+    # write before the second write on any putter is serviced.
+    for _ in range(NUM_PUTS):
+        remaining = set(range(NUM_PUTTERS))
+        for _ in range(NUM_PUTTERS):
+            v = await q.get()
+            assert v in remaining, "Unfair scheduling occured"
```
Comment (cmarqu) [inline tests/test_cases/test_cocotb/test_queues.py:179]:
```suggestion
            assert v in remaining, "Unfair scheduling occurred"
```

**Review by cmarqu — APPROVED** [review]
Just these tiny fixes...

File: cocotb/queue.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        if self._maxsize <= 0:
+            return False
+        else:
+            return self.qsize() >= self._maxsize
+
+    async def put(self, item: T) -> None:
+        """Put an *item* into the queue.
+
+        If the queue is full, wait until a free
+        slot is available before adding item.
```
Comment (cmarqu) [inline cocotb/queue.py:103]:
```suggestion
        slot is available before adding the item.
```

**Review by cmarqu — COMMENTED** [review]
Two more minor things from looking at the asyncio docs (which don't seem to be overly consistent in the markup).

**Comment by ktbarrett** [comment]
@cmarqu Documentation was from the asyncio docstrings. Most of the documentation issues exist there as well. https://github.com/python/cpython/blob/master/Lib/asyncio/queues.py

File: cocotb/queue.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    def __str__(self):
+        return '<{} {}>'.format(type(self).__name__, self._format())
+
+    def __class_getitem__(cls, type):
+        return cls
+
+    def _format(self):
+        result = 'maxsize={}'.format(repr(self._maxsize))
+        if getattr(self, '_queue', None):
+            result += ' _queue={}'.format(repr(list(self._queue)))
```
Comment (ktbarrett) [inline cocotb/queue.py:65]: I'm not entirely sure how this line isn't being covered. The base implementation has a `_queue` attribute.

**Review by ktbarrett — COMMENTED** [review]
Is there any objection to me removing the task pool features? `task_done`, `join`, and `_unfinished_tasks`? I mentioned how I seriously doubt they are going to be useful in a cocotb context before, and it's manual and easy to screw up as well, and there is no test for it.

**Comment by alexforencich** [comment]
My thought would be to keep them so that it's directly compatible with the asyncio queue.  But I suppose if the cocotb queue implements a subset of the functionality, then there should be no problem with substituting the asyncio queue if/when the scheduler supports asyncio code.

**Comment by alexforencich** [comment]
And why is "fixes to timer docstring" in here?

**Comment by ktbarrett** [comment]
@alexforencich I noticed it and didn't want to create a separate PR for a 1 line fix. Now its 4 lines, but things get away from you 🤷.

File: cocotb/queue.py (hunk 7 above)
Comment (marlonjames) [reply cocotb/queue.py:65]: I think it's because the `_queue` is empty at that point in the test, so the `if` is `False`.

**Comment by marlonjames** [comment]
I would be fine with removing the task tracking functionality. The basic features (and Exceptions) are compatible with `asyncio.Queue`, and we wouldn't include the task stuff if it wasn't in `asyncio`. Make it useful for cocotb but compatible where it makes sense.

**Comment by eric-wieser** [comment]
I don't really understand the argument for removing the task features - if they're useful in asyncio, why couldn't they also be useful in cocotb?

**Comment by ktbarrett** [comment]
1. There isn't a practical use case in a testbench. At least I can't think of one that isn't terribly contrived.
2. Breaks SRP.
3. It's not a very good implementation (must manually call `task_done` to signal a consumer has finished and anyone can do it, consumer must correctly handle possibility of exceptions to prevent it causing deadlock, there is no mapping of a particular input to task finishing, it's easy for beginners to misuse it, I could go on...).

If it's actually useful, we can add first class support for the functionality that avoids a lot of the design issues, and not add some hacked on afterthought that was obviously designed with the old threading system in mind and has been propagated since the early 00's.

File: cocotb/queue.py (hunk 3 above)
Comment (ktbarrett) [reply cocotb/queue.py:80]: I don't think that would have an affect on the value returned, so I'm sure it needs to be mentioned.

**Comment by marlonjames** [comment]
Should `__repr__` use `cocotb.triggers._pointer_str` ?

**Comment by ktbarrett** [comment]
@marlonjames It could. Do you have a recommendation?

File: cocotb/queue.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from typing import Generic, TypeVar
+import collections
+import heapq
+from asyncio import QueueEmpty, QueueFull
+
+import cocotb
+from cocotb.triggers import Event
```
Comment (marlonjames) [inline cocotb/queue.py:10]:
```suggestion
from cocotb.triggers import Event, _pointer_str
```

File: cocotb/queue.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+    def _wakeup_next(self, waiters):
+        while waiters:
+            event, task = waiters.popleft()
+            if not task._finished:
+                event.set()
+                break
+
+    def __repr__(self):
+        return '<{} at {:#x} {}>'.format(type(self).__name__, id(self), self._format())
```
Comment (marlonjames) [inline cocotb/queue.py:52]:
```suggestion
        return '<{} {} at {}>'.format(type(self).__name__, self._format(), _pointer_str(self))
```

File: tests/test_cases/test_cocotb/test_queues.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    q = Queue[int](maxsize=1)
+
+    q.put_nowait(0)
+    putter = cocotb.fork(q.put(1))
+
+    s = repr(q)
+    assert "maxsize" in s
+    assert "_queue" in s
+    assert "_putters" in s
+
```
Comment (marlonjames) [inline tests/test_cases/test_cocotb/test_queues.py:244]:
```suggestion
    assert str(q)[:-1] in s

```

File: tests/test_cases/test_cocotb/test_queues.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    assert q.get_nowait() == 0
+    getter = cocotb.fork(q.get())
+
+    s = repr(q)
+    assert "_putters" not in s
+    assert "_getters" in s
+
```
Comment (marlonjames) [inline tests/test_cases/test_cocotb/test_queues.py:251]:
```suggestion
    assert str(q)[:-1] in s

```

File: tests/test_cases/test_cocotb/test_queues.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+    await getter
+
+    s = repr(q)
+    assert "_getters" not in s
```
Comment (marlonjames) [inline tests/test_cases/test_cocotb/test_queues.py:255]:
```suggestion
    assert "_getters" not in s
    assert str(q)[:-1] in s
```

**Review by marlonjames — APPROVED** [review]
Suggested change to match repr format of triggers, and get the last bit of coverage.

File: tests/test_cases/test_cocotb/test_queues.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+    assert ref_q.qsize() == 0
+
+
+factory = TestFactory(run_queue_blocking_test)
+factory.add_option("queue_type", [Queue, PriorityQueue, LifoQueue])
+factory.generate_tests()
+
+
+@cocotb.test()
+async def test_repr(_):
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_queues.py:234]:
```suggestion
async def test_str_and_repr(_):
```

**Comment by ktbarrett** [comment]
Merging tomorrow if there are no further comments.

**Pull request merged**